### DISCUSSION TITLE
Auto Validator

### DISCUSSION THREAD

**Discussion by NourEldin-Ali** [imixs/open-bpmn#334] [Q&A]
Hi,

I'm working on auto validator.

I saw that in GLSP 2.x it was directly implemented. How we can add this functionality in your version? How we can access the model on submit (on modify any thing) to make validator?

Could you please help me to do that?
Thank you!

**Comment by rsoika** [comment]
Hi @NourEldin-Ali,
in Open-BPMN we bind the class `BPMNGLSPValidator` simply to our [BPMNDiagramModule](https://github.com/imixs/open-bpmn/blob/bf56ea4ff984bc8b48225f2db8591558dedb67c6/open-bpmn.glsp-server/src/main/java/org/openbpmn/glsp/BPMNDiagramModule.java#L221):

```java
    @Override
    protected Class<? extends ModelValidator> bindModelValidator() {
        return BPMNGLSPValidator.class;
    }
```

The validation itself is done in our case in the [Open-BPMN Meta model](https://github.com/imixs/open-bpmn/tree/master/open-bpmn.metamodel).  This is completely decoupled from the GLSP part and can be used on any BPMN model file.

We do not yet validate a lot of rules in BPMN , but the mechanism can be extended easily and I will extend this in the future. 

So for Open-BPMN you can simply call `validate` on any bpmn mode which returns a list of errors:


        List<BPMNValidationError> errorList = modelState.getBpmnModel().validate();


Does this answer your question?

**Reply by NourEldin-Ali** [reply]
Thank you for your reply!

I was already use the Meta model  to validate some part, however, we can't do everything using meta model. There for define some rules in my model in the validator [BPMNGLSPValidator](https://github.com/imixs/open-bpmn/blob/master/open-bpmn.glsp-server/src/main/java/org/openbpmn/glsp/validators/BPMNGLSPValidator.java), that can help me to detect other error or recommandations for the user.

For example, in my case, I'm working on extention of BPMN to add the data within the activity (it's a part of my PhD research). So, I need to detect the data that not connect to other data (see figure bellow).
![image](https://github.com/imixs/open-bpmn/assets/61513661/ae8f0fe2-910e-4db9-8da9-ca66bf57d435)

But, I need to click to thevalidation button in the palette to show the errors, and warning on the data elements.
![image](https://github.com/imixs/open-bpmn/assets/61513661/9219767d-0190-4ac8-ae21-6b3be395eb61)

My Question is: **how I can do that without click on the button?** I prefere to have this validation after add any elements to the model, but I dont know how I can access onSubmitModel/onUpdateModel or a function that called on modification.

**Reply by rsoika** [reply]
Hi  @NourEldin-Ali , ok I understand. I think it should be easy to send an action - the `RequestEditValidationAction` could be the thing you need. Look here: https://eclipse.dev/glsp/documentation/protocol/ 

On the other hand we should extend the `BPMNElementExtension` with a new method like `validate(BPMNElement bpmnElement)` . This may be useful if an adopter need additional validation rules. 

What do you think?

**Reply by chrisesposito4** [reply]
Hi, I've also got a question about validation. I can see the 'Validate model' button in the palette toolbar. I also have 2 send tasks with red 'x' markers.  Clicking the 'Validate model' button has no visible effect, but it occurs to me that maybe the red 'x's are the output of some model validation process. However, these markers appear when I load the file and before or without clicking the button at all. Is this button a placeholder for additional future functionality?

**Reply by rsoika** [reply]
Yes this button is a function to explizit call the validation process. In Open-BPMN we are currently validating on the fly. So normally this button bring no new effects. But it is necessary, because an adopter may implement additional validation logic that is not direct coupled into the GLSP-Object-Life-Cycle. And in this case this will be the option to re-validate the model.
